C-Plus-Plus/stack/string_expression_evaluation_bodmas.cpp

```diff
@@ -0,0 +1,203 @@
+int precedence(char c){
+	if(c=='*' || c=='/'){
+		return 3;
+	}
+	else if(c=='+' || c=='-'){
+		return 2;
+	}
+	else{
+		return 1;
+	}
+}
+
+//calculate function is used to calculate a value given two operands
+//and one operator
+int calculate(int num1, int num2, char symbol){
+	if(symbol=='*'){
+		return num1*num2;
+	}
+	else if(symbol=='/'){
+		return num1/num2;
+	}
+	else if(symbol=='+'){
+		return num1+num2;
+	}
+	else if(symbol=='-'){
+		return num1-num2;
+	}
+	return -1;
+}
+
+//Two stacks are used. 'values' stack is used to store the operands
+//'operators' stack is used to hold the operators
+int evaluate(string token){
+	stack<int>values;
+	stack<char>operators;
+	int i;
+	
+	//get length of given expression
+	int n = token.length();
+	
+	for(i=0; i<n; i++){
+		if(isdigit(token[i])){//digit encountered
+			int value = 0;
+			while(i<n && isdigit(token[i])){//there may be more than one digit number e.g(48+5) '48' has 2 digits
+				value = value*10 + (token[i]-'0');
+				i++;
+			}
+			
+			values.push(value);
+			i--;//decrement i to avoid incrementing it 2 times as outer for loop also increments it
+	
+		}
+		//if opening paranthesis is encontered simply push it in 'operators' stack
+		else if(token[i]=='('){
+			operators.push('(');
+		}
+
+		//If a closing paranthesis is encountered we have to evaluate the expression until an opening paranthesis is encountered
+		//Assuming the expression is valid we are guaranteed to obtain an opening paranthesis '('
+		else if(token[i]==')'){
+			while(!operators.empty() && operators.top()!='('){
+				//since stack stores values in Last In First Out(LIFO) fashion the top value in the 'values' stack is the second operand 
+				//e.g if expression is (4+5), the top value in stack will be 5 which is the second operand and 4 is the first operand
+				int val2 = values.top();
+				values.pop();
+
+				int val1 = values.top();
+				values.pop();
+
+				//get the operator from operator stack
+				char symbol = operators.top();
+				operators.pop();
+
+				//we now have 2 operands and one operator therefore we can calculate the value and store the calculated value back into 
+				//the 'values' stack for futher evaluation
+				int calculated_value = calculate(val1, val2, symbol);
+
+				values.push(calculated_value);
+			}
+
+			//pop opening brace as it is evaluated
+			operators.pop();
+		}
+		
+		else{//operator encountered
+			//if 'operators' stack is empty simply push the operator
+			if(operators.empty()){
+				operators.push(token[i]);
+			}
+			else{
+				//check precedence
+				//During expression evaluation NOTE that a strictly higher precedence operator can 
+				//sit on top of the current operator in 'operators' stack
+				//otherwise we need to evaluate the expression until the precedence of current operator in 'operators' stack becomes  
+				//strictly lower or 'operators' stack becomes empty
+				//if precedence of current symbol is more than top of 'operators' stack then simply push it
```

Comment:
Try to make Comments compact and easy to understand.

Reply:
updated